File: square/src/main/java/com/raxdenstudios/square/SquareMultiDexApplication.java
package com.raxdenstudios.square;

import android.content.res.Configuration;

import com.raxdenstudios.square.interceptor.ApplicationInterceptor;
import com.raxdenstudios.square.interceptor.Interceptor;
import com.raxdenstudios.square.manager.ApplicationMultiDexInterceptorManager;
import com.raxdenstudios.square.manager.InterceptorManagerFactory;

import java.util.ArrayList;
import java.util.List;

import androidx.multidex.MultiDexApplication;

/**
 * Created by Ángel Gómez
 *
 * SquareMultiDexApplication is an abstract class that adds interceptor functionality to the
 * MultiDexApplication.
 */
public abstract class SquareMultiDexApplication extends MultiDexApplication {

    private ApplicationMultiDexInterceptorManager mInterceptorManager;

    @Override
    public void onConfigurationChanged(Configuration newConfig) {
        super.onConfigurationChanged(newConfig);
        getInterceptorManager().onConfigurationChanged(newConfig);
    }

    @Override
    public void onCreate() {
        super.onCreate();
        getInterceptorManager().onCreate();
    }

    /* Support methods */

    protected abstract void setupInterceptors(List<Interceptor> interceptorList);

    private ApplicationMultiDexInterceptorManager getInterceptorManager() {
        if (mInterceptorManager == null) {
            mInterceptorManager = (ApplicationMultiDexInterceptorManager) InterceptorManagerFactory.buildManager(this);
            List<Interceptor> interceptorList = new ArrayList<>();
            setupInterceptors(interceptorList);
            for (Interceptor interceptor : interceptorList) {
                if (interceptor instanceof ApplicationInterceptor) {
                    mInterceptorManager.addInterceptor((ApplicationInterceptor) interceptor);
                }
            }
        }
        return mInterceptorManager;
    }

}

File: square/src/main/java/com/raxdenstudios/square/SquareMultiDexApplication.kt
package com.raxdenstudios.square

import android.content.res.Configuration
import android.support.multidex.MultiDexApplication
import com.raxdenstudios.square.interceptor.ApplicationInterceptor
import com.raxdenstudios.square.interceptor.Interceptor
import com.raxdenstudios.square.manager.ApplicationMultiDexInterceptorManager
import com.raxdenstudios.square.manager.InterceptorManagerFactory

/**
 * Created by Ángel Gómez
 *
 * SquareMultiDexApplication is an abstract class that adds interceptor functionality to the
 * MultiDexApplication.
 */
abstract class SquareMultiDexApplication : MultiDexApplication() {

    private var applicationMultiDexInterceptorManager: ApplicationMultiDexInterceptorManager? = null

    override fun onConfigurationChanged(newConfig: Configuration?) {
        super.onConfigurationChanged(newConfig)
        interceptorManager.onConfigurationChanged(newConfig)
    }

    override fun onCreate() {
        super.onCreate()
        interceptorManager.onCreate()
    }

    /* Support methods */

    protected abstract fun setupInterceptors(interceptorList: MutableList<Interceptor>)

    private val interceptorManager: ApplicationMultiDexInterceptorManager
        get() {
            return applicationMultiDexInterceptorManager ?: initInterceptorManager()
        }

    private fun initInterceptorManager(): ApplicationMultiDexInterceptorManager {
        applicationMultiDexInterceptorManager = InterceptorManagerFactory.buildManager(this) as ApplicationMultiDexInterceptorManager
        applicationMultiDexInterceptorManager?.apply {
            val interceptorList = mutableListOf<Interceptor>()
            setupInterceptors(interceptorList)
            interceptorList.forEach { interceptor -> addInterceptor((interceptor as ApplicationInterceptor<*>)) }
        }
        return applicationMultiDexInterceptorManager!!
    }

}
